software/apps/mandel-full/main.c

```diff
@@ -0,0 +1,210 @@
+    multicore_fifo_push_blocking(0);
+    while (!fractal.done && queue_get_level(&dvi0.q_tmds_valid) >= 5) generate_steal_one(&fractal);
+  }
+  __builtin_unreachable();
+}
+
+int __not_in_flash("main") main() {
+  vreg_set_voltage(VREG_VSEL);
+  sleep_ms(10);
+  set_sys_clock_khz(DVI_TIMING.bit_clk_khz, true);
+
+  setup_default_uart();
+
+  gpio_init(LED_PIN);
+  gpio_set_dir(LED_PIN, GPIO_OUT);
+  // gpio_put(LED_PIN, 1);
+  
+  init_palette();
+  init_mandel();
+
+  printf("Configuring DVI\n");
+
+  dvi0.timing = &DVI_TIMING;
+  dvi0.ser_cfg = DEFAULT_DVI_SERIAL_CONFIG;
+  dvi_init(&dvi0, next_striped_spin_lock_num(), next_striped_spin_lock_num());
+
+  printf("Core 1 start\n");
+  sem_init(&dvi_start_sem, 0, 1);
+  hw_set_bits(&bus_ctrl_hw->priority, BUSCTRL_BUS_PRIORITY_PROC1_BITS);
+  multicore_launch_core1(core1_main);
+
+  uint heartbeat = 0;
+  uint32_t encode_time = 0;
+
+  sem_release(&dvi_start_sem);
+  while (1) {
+    if (++heartbeat >= 30) {
+      heartbeat = 0;
+      gpio_xor_mask(1u << LED_PIN);
+
+      printf("Encode total time: %ldus\n", encode_time);
+      encode_time = 0;
+    }
+    if (fractal.done) zoom_mandel();
+    //if (heartbeat & 1) init_palette();
+    for (int y = 0; y < FRAME_HEIGHT / 2; y += 2) {
+      uint32_t *our_tmds_buf, *their_tmds_buf;
+      queue_remove_blocking_u32(&dvi0.q_tmds_free, &their_tmds_buf);
+      multicore_fifo_push_blocking((uint32_t)(&mandel[y*FRAME_WIDTH]));
+      multicore_fifo_push_blocking((uint32_t)their_tmds_buf);
+  
+      queue_remove_blocking_u32(&dvi0.q_tmds_free, &our_tmds_buf);
+      absolute_time_t start_time = get_absolute_time();
+      tmds_encode_palette_data((const uint32_t*)(&mandel[(y+1)*FRAME_WIDTH]), tmds_palette, our_tmds_buf, FRAME_WIDTH, PALETTE_BITS);
+      encode_time += absolute_time_diff_us(start_time, get_absolute_time());
+      
+      multicore_fifo_pop_blocking();
+
+      while (!fractal.done && queue_get_level(&dvi0.q_tmds_valid) >= 5) generate_one_forward(&fractal);
+
+      queue_add_blocking_u32(&dvi0.q_tmds_valid, &their_tmds_buf);
+      queue_add_blocking_u32(&dvi0.q_tmds_valid, &our_tmds_buf);
+    }
+    for (int y = FRAME_HEIGHT / 2 - 2; y >= 0; y -= 2) {
+      uint32_t *our_tmds_buf, *their_tmds_buf;
+      queue_remove_blocking_u32(&dvi0.q_tmds_free, &their_tmds_buf);
+      multicore_fifo_push_blocking((uint32_t)(&mandel[(y+1)*FRAME_WIDTH]));
+      multicore_fifo_push_blocking((uint32_t)their_tmds_buf);
+  
+      queue_remove_blocking_u32(&dvi0.q_tmds_free, &our_tmds_buf);
+      absolute_time_t start_time = get_absolute_time();
+      tmds_encode_palette_data((const uint32_t*)(&mandel[y*FRAME_WIDTH]), tmds_palette, our_tmds_buf, FRAME_WIDTH, PALETTE_BITS);
+      encode_time += absolute_time_diff_us(start_time, get_absolute_time());
+      
+      multicore_fifo_pop_blocking();
+
+      while (!fractal.done && queue_get_level(&dvi0.q_tmds_valid) >= 5) generate_one_forward(&fractal);
```

Comment:
Ah it's based on buffer health, I was wondering how you were doing this. I guess TMDS buffers are cheaper than they used to be :)